src/lib/logstore/log_store.cpp

```diff
@@ -375,6 +380,34 @@ logstore_seq_num_t HomeLogStore::get_contiguous_completed_seq_num(logstore_seq_n
     return (logstore_seq_num_t)m_records.completed_upto(from + 1);
 }
 
+void HomeLogStore::flush_sync(logstore_seq_num_t upto_seq_num) {
+    if (upto_seq_num == invalid_lsn()) { upto_seq_num = m_records.active_upto(); }
+
+    // if we have flushed already, we are done
+    if (m_records.completed_upto() >= upto_seq_num) { return; }
+
+    {
+        std::unique_lock lk(m_sync_flush_mtx);
+
+        // Step 1: Mark the waiter lsn to the seqnum we wanted to wait for. The completion of every lsn checks
+        // for this and if this lsn is completed, will make a callback which signals the cv.
+        m_sync_flush_waiter_lsn.store(upto_seq_num);
+
+        // Step 2: After marking this lsn, we again do a check, to avoid a race where completion checked for no lsn
+        // and the lsn is stored in step 1 above.
+        if (m_records.completed_upto() >= upto_seq_num) { return; }
```

Comment:
Thinking it more, we might not even need this check. 
Line 404 would just unblock wait since `m_records.completed_upto() >= upto_seq_num;`.
Or we are doing so to save calling `flush_if_needed`?

Reply:
Yes it is to save flush_if_needed call.